### PR TITLE
Upgrade `Python` 3.11 

### PR DESCRIPTION
**Pull request by bjornjorgensen**
## Describe your changes
Upgrade python from 3.10 to 3.11

[Release notes](https://www.python.org/downloads/release/python-3111/)

TensorFlow is now set to be installed using pip instead of mamba.

## Checklist (especially for first-time contributors)

- [ ] I have performed a self-review of my code
- [ ] If it is a core feature, I have added thorough tests
- [x] I will try not to use force-push to make the review process easier for reviewers
- [ ] I have updated the documentation for significant changes

### CODE REVIEW

**Comment by bjornjorgensen** [comment]
[3/7] RUN mamba install --quiet --yes     'altair'     'beautifulsoup4'     'bokeh'     'bottleneck'     'cloudpickle'     'conda-forge::blas=*=openblas'     'cython'     'dask'     'dill'     'h5py'     'ipympl'    'ipywidgets'     'matplotlib-base'     'numba'     'numexpr'     'openpyxl'     'pandas'     'patsy'     'protobuf'     'pytables'     'scikit-image'     'scikit-learn'     'scipy'     'seaborn'     'sqlalchemy'     'statsmodels'     'sympy'     'widgetsnbextension'    'xlrd' &&     mamba clean --all -f -y &&     fix-permissions "/opt/conda" &&     fix-permissions "/home/jovyan"
#6 sha256:ac03e34fe6e321c9e69e2e65b043b1b901e58cafc210ede3cd3df4959ad7cf45
#6 32.22 Could not solve for environment specs
#6 32.22 Encountered problems while solving:
#6 32.22   - nothing provides openssl >=1.0.2p,<1.0.3a needed by python-3.6.7-hd21baee_1002
#6 32.22 
#6 32.22 The environment can't be solved, aborting the operation
#6 32.22 



[3/7] RUN mamba install --quiet --yes     'altair'     'beautifulsoup4'     'bokeh'     'bottleneck'     'cloudpickle'     'conda-forge::blas=*=openblas'     'cython'     'dask'     'dill'     'h5py'     'ipympl'    'ipywidgets'     'matplotlib-base'     'numba'     'numexpr'     'openpyxl'     'pandas'     'patsy'     'protobuf'     'pytables'     'scikit-image'     'scikit-learn'     'scipy'     'seaborn'     'sqlalchemy'     'statsmodels'     'sympy'     'widgetsnbextension'    'xlrd' &&     mamba clean --all -f -y &&     fix-permissions "/opt/conda" &&     fix-permissions "/home/jovyan"
#6 sha256:481075b6d100da5c3b645b3cd2f6cb3b7edd0aee0c3749d4174a2daa1c3eaf2c
#6 77.07 Could not solve for environment specs
#6 77.07 Encountered problems while solving:
#6 77.07   - package numba-0.53.0-py39h4292f66_0 requires python >=3.9,<3.10.0a0, but none of the providers can be installed
#6 77.07 
#6 77.07 The environment can't be solved, aborting the operation
#6 77.07

**Comment by bjornjorgensen** [comment]
numba https://github.com/numba/numba/issues/8304

**Comment by bjornjorgensen** [comment]
I think we should wait for spark 3.4 before we upgrade 
https://github.com/apache/spark/pull/38987
Spark 3.4 will be out in jan / feb.

**Comment by bjornjorgensen** [comment]
I can test that when https://github.com/jupyter/docker-stacks/pull/1845 is merged.

**Comment by mathbunnyru** [comment]
> I can test that when #1845 is merged.

Haha, I guess my PR came right on time 😂

**Comment by benz0li** [comment]
> Do other packages work properly and have you tried removing/commenting `numba` and `pyspark` and building/testing everything else?

@mathbunnyru Regarding `scipy-notebook`: The current release of `pytables` does not support Python 3.11.  
ℹ️ See also https://github.com/jupyter/docker-stacks/pull/1812#issuecomment-1290946139.

**Comment by parmentelat** [comment]
@ benz0li
> Regarding scipy-notebook: The current release of pytables does not support Python 3.11.

not sure if this is still current or not, the `tables` package seems to be pip-installable on 3.11 no ?

**Comment by benz0li** [comment]
> not sure if this is still current or not, the `tables` package seems to be pip-installable on 3.11 no ?

@parmentelat See https://github.com/jupyter/docker-stacks/pull/1812#issuecomment-1290946139. There is only one issue remaining: https://github.com/numba/numba/issues/8304.

**Comment by mathbunnyru** [comment]
I merge main to this PR and removed `numba` to make sure everything else works as expected.

**Comment by bjornjorgensen** [comment]
tensorflow doesn't support python 3.11 https://stackoverflow.com/questions/74556733/tensorflow-support-for-python3-11

**Comment by mathbunnyru** [comment]
Also [`test_nbconvert` fails](https://github.com/jupyter/docker-stacks/actions/runs/3966152752/jobs/6797478147), so we will probably have to take a look there as well.

**Comment by bjornjorgensen** [comment]
Apache spark will have a rc1 next week. Spark 3.4 have support for python 3.11 I can make a test run with Spark 3.4 rc1 and python 3.11

**Comment by bjornjorgensen** [comment]
Spark 3.3.2 does not have support for python 3.11.
Spark 3.4 will soon have a RC1 witch have support for python 3.11

**Comment by mathbunnyru** [comment]
Numba and TensorFlow issues about python 3.11 were closed.
Now we need to wait until the recipes in conda-forge are updated.

**Comment by bjornjorgensen** [comment]
great, but Apache spark 3.4 will have another RC. So I think we need to wait one or two weeks.

**Comment by benz0li** [comment]
> Numba and TensorFlow issues about python 3.11 were closed.

See the Release Candidate checklist for Numba 0.57.0rc1 at https://github.com/numba/numba/issues/8841.

**Comment by benz0li** [comment]
[Numba v0.57.0](https://github.com/numba/numba/releases/tag/0.57.0) was released yesterday:

> A new release of Numba adding support for Python 3.11, NumPy 1.24, and upgrading to LLVM 14. For details, please see the change log at https://numba.readthedocs.io/en/stable/release-notes.html#version-0-57-0-1-may-2023.

**Comment by mathbunnyru** [comment]
> [Numba v0.57.0](https://github.com/numba/numba/releases/tag/0.57.0) was released yesterday:
> 
> > A new release of Numba adding support for Python 3.11, NumPy 1.24, and upgrading to LLVM 14. For details, please see the change log at https://numba.readthedocs.io/en/stable/release-notes.html#version-0-57-0-1-may-2023.

Great news 👍 We will have to wait for conda-forge release, hope it won't take too long.
https://github.com/conda-forge/numba-feedstock/pull/115

**Comment by mathbunnyru** [comment]
> > [Numba v0.57.0](https://github.com/numba/numba/releases/tag/0.57.0) was released yesterday:
> > > A new release of Numba adding support for Python 3.11, NumPy 1.24, and upgrading to LLVM 14. For details, please see the change log at https://numba.readthedocs.io/en/stable/release-notes.html#version-0-57-0-1-may-2023.
> 
> Great news 👍 We will have to wait for conda-forge release, hope it won't take too long. [conda-forge/numba-feedstock#115](https://github.com/conda-forge/numba-feedstock/pull/115)

This PR has been just merge :tada:

**Comment by bjornjorgensen** [comment]
Tensorflow version 2.12.0 have support for python 3.11 https://github.com/tensorflow/tensorflow/releases/tag/v2.12.0 
But the "Do not install TensorFlow with conda. It may not have the latest stable version. pip is recommended since TensorFlow is only officially released to PyPI."
https://www.tensorflow.org/install/pip

**Comment by bjornjorgensen** [comment]
should we mamba insall pip in https://github.com/jupyter/docker-stacks/blob/main/tensorflow-notebook/Dockerfile
and then pip install tensorflow?

**Comment by mathbunnyru** [comment]
In my experience, it's always better to use conda/mamba and not mix them with pip.
There is active work on the new TensorFlow here: https://github.com/conda-forge/tensorflow-feedstock/pull/316
So, let's wait for a while, I think they might release new version quite soon.

**Comment by njzjz** [comment]
I don't find a workaround to resolve this error:
```
ERROR: /home/conda/feedstock_root/build_artifacts/tensorflow-split_1684490216303/work/tensorflow/c/BUILD:359:11: Compiling tensorflow/c/tf_status.cc failed: (Exit 1): x86_64-conda-linux-gnu-cc failed: error executing command 
  (cd /home/conda/feedstock_root/build_artifacts/tensorflow-split_1684490216303/_build_env/share/bazel/07f14a670c6615d0600f0f6296f25be0/execroot/org_tensorflow && \
  exec env - \
    PATH=/home/conda/feedstock_root/build_artifacts/tensorflow-split_1684490216303/work:/home/conda/feedstock_root/build_artifacts/tensorflow-split_1684490216303/_build_env/bin:/home/conda/feedstock_root/build_artifacts/tensorflow-split_1684490216303/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_plac/bin:/opt/conda/condabin:/home/conda/feedstock_root/build_artifacts/tensorflow-split_1684490216303/_build_env/bin:/home/conda/feedstock_root/build_artifacts/tensorflow-split_1684490216303/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_plac/bin:/opt/conda/bin:/opt/conda/condabin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/home/conda/bin \
    PWD=/proc/self/cwd \
    PYTHON_BIN_PATH=/home/conda/feedstock_root/build_artifacts/tensorflow-split_1684490216303/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_plac/bin/python \
    PYTHON_LIB_PATH=/home/conda/feedstock_root/build_artifacts/tensorflow-split_1684490216303/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_plac/lib/python3.10/site-packages \
    TF2_BEHAVIOR=1 \
    TF_SYSTEM_LIBS=absl_py,astor_archive,astunparse_archive,boringssl,com_github_googlecloudplatform_google_cloud_cpp,com_github_grpc_grpc,com_google_protobuf,curl,cython,dill_archive,flatbuffers,gast_archive,gif,icu,libjpeg_turbo,org_sqlite,png,pybind11,snappy,zlib \
  /home/conda/feedstock_root/build_artifacts/tensorflow-split_1684490216303/_build_env/bin/x86_64-conda-linux-gnu-cc -U_FORTIFY_SOURCE -fstack-protector -Wall -Wunused-but-set-parameter -Wno-free-nonheap-object -fno-omit-frame-pointer -g0 -O2 '-D_FORTIFY_SOURCE=1' -DNDEBUG -ffunction-sections -fdata-sections '-std=c++0x' -MD -MF bazel-out/k8-opt/bin/tensorflow/c/_objs/tf_status/tf_status.pic.d '-frandom-seed=bazel-out/k8-opt/bin/tensorflow/c/_objs/tf_status/tf_status.pic.o' -fPIC -DEIGEN_MPL2_ONLY '-DEIGEN_MAX_ALIGN_BYTES=64' -iquote . -iquote bazel-out/k8-opt/bin -iquote external/eigen_archive -iquote bazel-out/k8-opt/bin/external/eigen_archive -iquote external/com_google_absl -iquote bazel-out/k8-opt/bin/external/com_google_absl -iquote external/nsync -iquote bazel-out/k8-opt/bin/external/nsync -iquote external/double_conversion -iquote bazel-out/k8-opt/bin/external/double_conversion -iquote external/com_google_protobuf -iquote bazel-out/k8-opt/bin/external/com_google_protobuf -iquote external/gif -iquote bazel-out/k8-opt/bin/external/gif -iquote external/libjpeg_turbo -iquote bazel-out/k8-opt/bin/external/libjpeg_turbo -iquote external/com_googlesource_code_re2 -iquote bazel-out/k8-opt/bin/external/com_googlesource_code_re2 -iquote external/farmhash_archive -iquote bazel-out/k8-opt/bin/external/farmhash_archive -iquote external/fft2d -iquote bazel-out/k8-opt/bin/external/fft2d -iquote external/highwayhash -iquote bazel-out/k8-opt/bin/external/highwayhash -iquote external/zlib -iquote bazel-out/k8-opt/bin/external/zlib -isystem third_party/eigen3/mkl_include -isystem bazel-out/k8-opt/bin/third_party/eigen3/mkl_include -isystem external/eigen_archive -isystem bazel-out/k8-opt/bin/external/eigen_archive -isystem external/nsync/public -isystem bazel-out/k8-opt/bin/external/nsync/public -isystem external/farmhash_archive/src -isystem bazel-out/k8-opt/bin/external/farmhash_archive/src -Wno-all -Wno-extra -Wno-deprecated -Wno-deprecated-declarations -Wno-ignored-attributes -Wno-array-bounds -Wunused-result '-Werror=unused-result' -Wswitch '-Werror=switch' '-Wno-error=unused-but-set-variable' -DAUTOLOAD_DYNAMIC_KERNELS '-std=c++17' -DEIGEN_AVOID_STL_ARRAY -Iexternal/gemmlowp -Wno-sign-compare '-ftemplate-depth=900' -fno-exceptions -DINTEL_MKL -DAMD_ZENDNN -msse3 -pthread -fno-canonical-system-headers -Wno-builtin-macro-redefined '-D__DATE__="redacted"' '-D__TIMESTAMP__="redacted"' '-D__TIME__="redacted"' -c tensorflow/c/tf_status.cc -o bazel-out/k8-opt/bin/tensorflow/c/_objs/tf_status/tf_status.pic.o)
# Configuration: 311988e8e1d0b6f908e732a2a226b7c4c672bd3781cfe926d0de72cee0e8a023
# Execution platform: @local_execution_config_platform//:platform
In file included from ./tensorflow/tsl/platform/status.h:37,
                 from ./tensorflow/tsl/c/tsl_status_internal.h:19,
                 from ./tensorflow/c/tf_status_internal.h:19,
                 from tensorflow/c/tf_status.cc:18:
bazel-out/k8-opt/bin/tensorflow/tsl/protobuf/error_codes.pb.h:10:10: fatal error: google/protobuf/port_def.inc: No such file or directory
   10 | #include <google/protobuf/port_def.inc>
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
```

**Comment by mathbunnyru** [comment]
@bjornjorgensen sorry for triggering CI so many times and making GitHub spam you - it was a really nice opportunity to test our new aarch64 runners 😂

**Comment by mathbunnyru** [comment]
Also, currently, I think we can proceed with pypi tensorflow.
Let's give it at least a try.

I also suppose, we will be able to build aarch64 `tensorflow-notebook`, but I suggest to enable it in a separate PR after merging this one.

**Comment by mathbunnyru** [comment]
> so you will merge this one now

No, please, let's try use pip in this PR and if it works, I will merge it.
About the errors - are you sure we will have these errors?
I suppose we're using non-coda version, so we shouldn't see there errors?

**Comment by bjornjorgensen** [comment]
> No, please, let's try use pip in this PR and if it works, I will merge it. About the errors - are you sure we will have these errors? I suppose we're using non-coda version, so we shouldn't see there errors?

I try to follow https://stackoverflow.com/questions/41060382/using-pip-to-install-packages-to-anaconda-environment 
so now I hope that we install pip in conda and use pip witch are in conda to install tensorflow.

**Comment by bjornjorgensen** [comment]
a local test 

```
# Copyright (c) Jupyter Development Team.
# Distributed under the terms of the Modified BSD License.
ARG OWNER=jupyter
ARG BASE_CONTAINER=$OWNER/scipy-notebook
FROM $BASE_CONTAINER

LABEL maintainer="Jupyter Project <jupyter@googlegroups.com>"

# Fix: https://github.com/hadolint/hadolint/wiki/DL4006
# Fix: https://github.com/koalaman/shellcheck/wiki/SC3014
SHELL ["/bin/bash", "-o", "pipefail", "-c"]



ARG PYTHON_VERSION=3.11

# Setup work directory for backward-compatibility
#RUN mkdir "/home/${NB_USER}/work" && \
#    fix-permissions "/home/${NB_USER}"
# Download and install Micromamba, and initialize Conda prefix.
#   <https://github.com/mamba-org/mamba#micromamba>
#   Similar projects using Micromamba:
#     - Micromamba-Docker: <https://github.com/mamba-org/micromamba-docker>
#     - repo2docker: <https://github.com/jupyterhub/repo2docker>
# Install Python, Mamba and jupyter_core
# Cleanup temporary files and remove Micromamba
# Correct permissions
# Do all this in a single RUN command to avoid duplicating all of the
# files across image layers when the permissions change
RUN rm -R "${CONDA_DIR}/conda-meta/pinned"
COPY --chown="${NB_UID}:${NB_GID}" initial-condarc "${CONDA_DIR}/.condarc"
WORKDIR /tmp
RUN set -x && \
    arch=$(uname -m) && \
    if [ "${arch}" = "x86_64" ]; then \
        # Should be simpler, see <https://github.com/mamba-org/mamba/issues/1437>
        arch="64"; \
    fi && \
    wget -qO /tmp/micromamba.tar.bz2 \
        "https://micromamba.snakepit.net/api/micromamba/linux-${arch}/latest" && \
    tar -xvjf /tmp/micromamba.tar.bz2 --strip-components=1 bin/micromamba && \
    rm /tmp/micromamba.tar.bz2 && \
    PYTHON_SPECIFIER="python=${PYTHON_VERSION}" && \
    if [[ "${PYTHON_VERSION}" == "default" ]]; then PYTHON_SPECIFIER="python"; fi && \
    # Install the packages
    ./micromamba install \
        --root-prefix="${CONDA_DIR}" \
        --prefix="${CONDA_DIR}" \
        --yes \
        "${PYTHON_SPECIFIER}" \
        'mamba' \
        'jupyter_core' && \
    rm micromamba && \
    # Pin major.minor version of python
    mamba list python | grep '^python ' | tr -s ' ' | cut -d ' ' -f 1,2 >> "${CONDA_DIR}/conda-meta/pinned" && \
    mamba clean --all -f -y && \
    fix-permissions "${CONDA_DIR}" && \
    fix-permissions "/home/${NB_USER}"


# Install Tensorflow with pip
RUN mamba install --yes \
    'pip' && \
    mamba clean --all -f -y && \
    pip install tensorflow && \
    fix-permissions "${CONDA_DIR}" && \
    fix-permissions "/home/${NB_USER}"
```
```
docker build --pull --rm -f "Dockerfile" -t testdocker:latest "." 
```


```

Sending build context to Docker daemon   5.12kB
Step 1/11 : ARG OWNER=jupyter
Step 2/11 : ARG BASE_CONTAINER=$OWNER/scipy-notebook
Step 3/11 : FROM $BASE_CONTAINER
latest: Pulling from jupyter/scipy-notebook
..........................................


Successfully installed absl-py-1.4.0 astunparse-1.6.3 cachetools-5.3.1 flatbuffers-23.5.26 gast-0.4.0 google-auth-2.19.0 google-auth-oauthlib-1.0.0 google-pasta-0.2.0 grpcio-1.54.2 jax-0.4.10 keras-2.12.0 libclang-16.0.0 markdown-3.4.3 ml-dtypes-0.1.0 numpy-1.23.5 opt-einsum-3.3.0 pyasn1-0.5.0 pyasn1-modules-0.3.0 requests-oauthlib-1.3.1 rsa-4.9 tensorboard-2.12.3 tensorboard-data-server-0.7.0 tensorflow-2.12.0 tensorflow-estimator-2.12.0 tensorflow-io-gcs-filesystem-0.32.0 termcolor-2.3.0 urllib3-1.26.16 werkzeug-2.3.4 wrapt-1.14.1
Removing intermediate container cc6653edfa0f
 ---> ab70dc9c1c01
Successfully built ab70dc9c1c01
Successfully tagged testdocker:latest

```

File: tensorflow-notebook/Dockerfile
Comment (mathbunnyru) [inline tensorflow-notebook/Dockerfile:16]: I've just checked `mamba list` and `pip` is already there, so we don't need to install it here, I suppose.

File: tensorflow-notebook/Dockerfile
Comment (bjornjorgensen) [inline tensorflow-notebook/Dockerfile:15]: Do we need this?

Comment (mathbunnyru) [reply tensorflow-notebook/Dockerfile:15]: Not sure (it was here before me), but we have it after every mamba/pip installation, so let's keep it.

**Review by mathbunnyru — APPROVED** [review]
I will merge when the tests pass.

**Comment by bjornjorgensen** [comment]
ok, should we update the PR description?

**Comment by mathbunnyru** [comment]
> ok, should we update the PR description?

I think it's ok now.

**Pull request merged**
